lib/libzpool/kernel.c

```diff
@@ -813,11 +819,37 @@ __dprintf(const char *file, const char *func, int line, const char *fmt, ...)
 			(void) printf("%llu ", gethrtime());
 		if (dprintf_find_string("long"))
 			(void) printf("%s, line %d: ", newfile, line);
-		(void) printf("%s: ", func);
+		(void) printf("dprintf: %s: ", func);
 		va_start(adx, fmt);
 		(void) vprintf(fmt, adx);
 		va_end(adx);
 		funlockfile(stdout);
+	} else {
+		/* zfs_dbgmsg is logged for dumping later */
+		size_t size;
+		char *buf, *nl;
+		int i;
+
+		size = 1024;
+		buf = umem_alloc(size, UMEM_NOFAIL);
+		i = snprintf(buf, size, "%s:%d:%s(): ", newfile, line, func);
+
+		if (i < size) {
+			va_start(adx, fmt);
+			(void) vsnprintf(buf + i, size - i, fmt, adx);
+			va_end(adx);
+		}
+
+		/*
+		 * Get rid of trailing newline.
+		 */
```

Comment:
why is this needed?  are we zfs_dbgmsg()-ing things that have newlines?  Seems like we should fix that rather than changing the string after the fact (in some places but not others).

Reply:
This was taken directly from the [zfs_debug.c code](https://github.com/zfsonlinux/zfs/blob/master/module/zfs/zfs_debug.c#L148-L208). I just wanted to ensure they behaved the same.

Reply:
Let's fix zfs_debug.c too then.

Reply:
Ah. I think I see why this was done. `dprintf()` strings (by convention it seems) have a trailing newline while `zfs_dbgmsg()` strings do not. After looking around, I don't see any examples of multi-line calls to either of these functions. Not sure what the best way to handle this is....

Reply:
Ok. I can do that.